fv3core/stencils/riem_solver_c.py

```diff
@@ -57,89 +53,113 @@ def finalize(
-    cp3 = utils.make_storage_from_shape(shape, riemorigin, cache_key="riemc_cp3")
-    w = utils.make_storage_from_shape(shape, riemorigin, cache_key="riemc_w")
-    pem = utils.make_storage_from_shape(
-        shape, riemorigin, cache_key="riem_solver_c_pem"
-    )
-    pe = utils.make_storage_from_shape(shape, riemorigin, cache_key="riem_solver_c_pe")
-    gm = utils.make_storage_from_shape(shape, riemorigin, cache_key="riem_solver_c_gm")
-    dz = utils.make_storage_from_shape(shape, riemorigin, cache_key="riem_solver_c_dz")
-    pm = utils.make_storage_from_shape(shape, riemorigin, cache_key="riem_solver_c_pm")
-    # it looks like this code sets pef = ptop, and does not otherwise use pef here
-    precompute(
-        delpc,
-        cappa,
-        w3,
-        w,
-        cp3,
-        gz,
-        dm,
-        q_con,
-        pem,
-        dz,
-        gm,
-        pm,
-        ptop,
-        origin=riemorigin,
-        domain=domain,
-    )
+class RiemannSolverC:
+    """
+    Fortran subroutine Riem_Solver_C
+    """
+
+    def __init__(self, namelist):
+        grid = spec.grid
+        origin = grid.compute_origin(add=(-1, -1, 0))
+        domain = grid.domain_shape_compute(add=(2, 2, 1))
+        shape = grid.domain_shape_full(add=(1, 1, 1))
+
+        self._dm = utils.make_storage_from_shape(shape, origin)
+        self._w = utils.make_storage_from_shape(shape, origin)
+        self._pem = utils.make_storage_from_shape(shape, origin)
+        self._pe = utils.make_storage_from_shape(shape, origin)
+        self._gm = utils.make_storage_from_shape(shape, origin)
+        self._dz = utils.make_storage_from_shape(shape, origin)
+        self._pm = utils.make_storage_from_shape(shape, origin)
+
+        self._precompute_stencil = StencilWrapper(
+            precompute,
+            origin=origin,
+            domain=domain,
+        )
+        self._sim1_solve = Sim1Solver(
+            namelist,
+            grid,
+            grid.is_ - 1,
+            grid.ie + 1,
+            grid.js - 1,
+            grid.je + 1,
+        )
+        self._finalize_stencil = StencilWrapper(
+            finalize,
+            origin=origin,
+            domain=domain,
+        )
 
-    sim1_solve = utils.cached_stencil_class(Sim1Solver)(
-        spec.namelist,
-        spec.grid,
-        is1,
-        ie1,
-        js1,
-        je1,
-        cache_key="riem_solver_c_sim1solver",
-    )
-    sim1_solve(dt2, gm, cp3, pe, dm, pm, pem, w, dz, ptc, ws)
+    def __call__(
+        self,
+        dt2: float,
+        cappa: FloatField,
+        ptop: float,
+        hs: FloatFieldIJ,
+        w3: FloatField,
+        ptc: FloatField,
+        q_con: FloatField,
+        delpc: FloatField,
+        gz: FloatField,
+        pef: FloatField,
+        ws: FloatFieldIJ,
+    ):
+        """
+        Solves for the nonhydrostatic terms for vertical velocity (w)
+        and non-hydrostatic pressure perturbation after D-grid winds advect
```

Comment:
This one is done after the C-grid winds are advected

Reply:
D'oh! Good catch, I meant to change that.